### PR TITLE
feat: allow bgworkers to access JWT claims

### LINKED ISSUE #6: Compute team review
- [ ] Remove the "Created by pgrx" from the extension control file.
- [ ] Why can't the extension be relocatable (control file)?
- [ ] I don't see any `pgrx_embed!()`. Is it important?
- [ ] Should we set version to something other than 0.0.0 for the initial inclusion into the compute image?
- [ ] Are the thread local variables supposed to be per-backend? Doesn't seem like they need to be thread local to me. I think this has been previously discussed.
- [ ] I don't think a panic is the correct way to do error handling, like in `auth::init()`, which can only be called once. In postgres, you would use an [ereport](https://docs.rs/pgrx/latest/pgrx/macro.ereport.html).
- [ ] Can we evaluate our unwrap/expect usage, and instead use ereport?
- [ ] Mind renaming `auth.get()` to something like `auth.get_session()`? Anything but a naked `get` would be good. Looking at `auth.user_id()` which is a getter, maybe something like `auth.session()` makes sense.
- [ ] Can we use a more descriptive comment in the control file, similar to [pgvector](https://github.com/pgvector/pgvector/blob/77b3d1f2a8e08aa6ca2545eddd9fb51796c32ea1/vector.control#L1)?
- [ ] Can we rename the schema from `auth` to `session_jwt` or something? `auth` is pretty generic and not related to the extension name.

I'll leave the Rust stuff to the experts :). It's possible pgrx is taking the panics and transforming them into Postgres errors, but if it does do that, what error messages does it produce for unwraps, and what SQL error code?

### PR DESCRIPTION
**Pull request by mrl5**
related to https://github.com/neondatabase/cloud/issues/16041
and this checkpoint from https://github.com/neondatabase/pg_session_jwt/issues/6
> Are the thread local variables supposed to be per-backend? Doesn't seem like they need to be thread local to me. I think this has been previously discussed.

closes #6

### CODE REVIEW

**Comment by mrl5** [comment]
GUC init influenced by https://github.com/pgcentralfoundation/pgrx/issues/128#issuecomment-826934395

File: src/gucs.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+pub static NEON_AUTH_JWT_RUNTIME_PARAM: &'static str = "neon.auth.jwt";
+pub static NEON_AUTH_JWT: GucSetting<Option<&'static CStr>> = GucSetting::<Option<&'static CStr>>::new(None);
+
+pub fn init() {
+    GucRegistry::define_string_guc(
+        &NEON_AUTH_JWK_RUNTIME_PARAM,
+        "JSON Web Key (JWK) userd for JWT validation",
+        "Generated per connection by Neon local proxy",
+        &NEON_AUTH_JWK,
+        GucContext::Suset, // we should use GucContext::SuBackend but this breaks unit tests
```
Comment (mrl5) [inline src/gucs.rs:15]: I still don't know:
- how to set it properly in unit tests
- how to set it properly while running `cargo pgrx run pg16`

so for now I'm using `Suset`

File: src/lib.rs [hunk 2]
```diff
@@ -186,6 +208,14 @@ pub mod auth {
     /// This function will panic if the JWT could not be verified.
     #[pg_extern]
     pub fn jwt_session_init(s: &str) {
+        Spi::run(format!("SET {} = '{}'", NEON_AUTH_JWT_RUNTIME_PARAM, s).as_str()).unwrap_or_else(|e| {
+            error_code!(
+                PgSqlErrorCode::ERRCODE_S_R_E_PROHIBITED_SQL_STATEMENT_ATTEMPTED,
```
Comment (mrl5) [inline src/lib.rs:213]: I'd appreciate some1 confirming or suggesting proper ERRCODE for this case

File: src/gucs.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+use pgrx::*;
+use std::ffi::CStr;
+
+pub static NEON_AUTH_KID_RUNTIME_PARAM: &str = "neon.auth.kid";
+pub static NEON_AUTH_KID: GucSetting<i32> = GucSetting::<i32>::new(0);
```
Comment (mrl5) [inline src/gucs.rs:5]: GUC doesn't support `i64`, later `i32` is casted into `i64` for backwards compatibility

**Comment by mrl5** [comment]
tested manually with:
```console
export PGOPTIONS='-c neon.auth.kid=1 -c neon.auth.jwk={"crv":"P-256","kty":"EC","x":"CV9leSCKaIvTHgFe1Anhl8ylOu73QmZfGQ7nibnAFpo","y":"8PZKd0_9EZ6-Qm_on5DdO2LErp3T4LCEs7gBZWj3ACs"} -c neon.auth.jwt=eyJraWQiOjF9.eyJzdWIiOiJrb3RiZWhlbW90IiwianRpIjozfQ.5anyLNZ-8w8LycJnhIke_j3UZI-lEKXek1JhFWxLYD71VulwnBjPoTFVXY5fW31i2khdeMCxOeYGB0C10H1w3w'
cargo pgrx run pg16
```
```sql
drop extension pg_session_jwt; create extension pg_session_jwt;
select auth.user_id();
```

File: src/lib.rs [hunk 4]
```diff
@@ -185,7 +215,35 @@ pub mod auth {
     ///
     /// This function will panic if the JWT could not be verified.
     #[pg_extern]
-    pub fn jwt_session_init(s: &str) {
+    pub fn jwt_session_init(jwt: &str) {
+        Spi::run(format!("SET {} = '{}'", NEON_AUTH_JWT_RUNTIME_PARAM, jwt).as_str())
+            .unwrap_or_else(|e| {
+                error_code!(
+                    PgSqlErrorCode::ERRCODE_S_R_E_PROHIBITED_SQL_STATEMENT_ATTEMPTED,
```
Comment (mrl5) [inline src/lib.rs:222]: I'd appreciate some1 confirming or suggesting proper ERRCODE for this case

File: src/gucs.rs (hunk 1 above)
Comment (conradludgate) [reply src/gucs.rs:30]: I can write a manual test suite then that opens a new connection reach time. Should not be too much of a bother

File: src/gucs.rs (hunk 3 above)
Comment (conradludgate) [reply src/gucs.rs]: In a pass I removed this GUC and folded it into the JWK

File: pgrx-tests/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# pgrx-tests
+
+Test framework for [`pgrx`](https://crates.io/crates/pgrx/).
+
+Meant to be used as one of your `[dev-dependencies]` when using `pgrx`.
+
+Forked off of pgrx 0.11.3 by Conrad Ludgate for the purposes of adding support for
+1. Providing options used for initialising GucContext::Backend
+2. Running tests as non-superuser
```
Comment (conradludgate) [inline pgrx-tests/README.md:9]: I will chat with the pgrx team and see what we can do here to upstream these ideas

File: tests/pg_session_jwt.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+    tx.execute(&format!("set neon.auth.jwt = '{jwt}'"), &[])?;
+    let user_id = tx.query_one("select auth.user_id()", &[])?;
+    let user_id = user_id.get::<_, String>("user_id");
+    assert_eq!(user_id, "foo");
+
+    Ok(())
+}
+
+// fn discard() -> eyre::Result<()> {
```
Comment (mrl5) [inline tests/pg_session_jwt.rs:156]: @conradludgate can I remove this commented function or you'd prefer to have it committed?

**Review by conradludgate — APPROVED** [review]

**Pull request merged**